### PR TITLE
Update state when defaultValue changes

### PR DESCRIPTION
**Pull request by adrienharnay**
Sometimes you want your component to not be controlled, but still be able to update the defaultValue externally.

I also added the same logic in componentDidUpdate than in componentDidMount, so that `props.value` and `props.defaultValue` changes would trigger number formatting and flag changes.

Please let me know if I should add/change anything

### CODE REVIEW

**Comment by adrienharnay** [comment]
Hey @patw0929 could you please tell me how you feel about this PR? I think I covered everything regarding the initial problematic (and have actually been relying on this fork for a couple months in production), but let me know if I should add/change anything. Thanks!

**Comment by adrienharnay** [comment]
Also, I adapted my changes to the updates made on master since I opened the PR

**Comment by patw0929** [comment]
Thank you @adrienharnay!
But I have to work overtime recently, just got home... 😫 I can review the PR at weekend.
Maybe @superhit0 can help you out?

**Comment by superhit0** [comment]
Thanks @adrienharnay @patw0929 , But I am also down with heavy work till this weekend.
Will be available only after this saturday.

@adrienharnay can you please explain on a high level what we are trying to do here?
Thanks

**Comment by adrienharnay** [comment]
Hey, no worries! Let's talk again this week-end :)

Basically, sometimes you are using TelInput in an uncontrolled way (not passing `props.value`), but provide `props.defaultValue`. When `props.defaultValue` changes, value should change accordingly in TelInput to reflect this change.

[react-final-form](https://github.com/final-form/react-final-form) does this with the `initialValues` parameter, for example.

**Comment by adrienharnay** [comment]
Hey, have you had any time to look at this PR? Thanks

File: src/components/IntlTelInputApp.js [hunk 1]
```diff
@@ -184,8 +178,12 @@ class IntlTelInputApp extends Component {
   }
 
   componentDidUpdate(prevProps) {
-    if (this.props.value !== prevProps.value) {
-      this.updateFlagFromNumber(this.props.value);
```
[the remark this answers is not in the document]
Comment (adrienharnay) [reply src/components/IntlTelInputApp.js]: `updateFlagFromNumber` will be called in `setInitialState` if needs be

File: src/components/IntlTelInputApp.js [hunk 2]
```diff
@@ -184,8 +178,12 @@ class IntlTelInputApp extends Component {
...
+    if (
```
[the remark this answers is not in the document]
Comment (adrienharnay) [reply src/components/IntlTelInputApp.js]: Calling `updateFlagFromNumber` was nice, but I think it was not enough, as we didn't call `updateValFromNumber` (which we know do)

File: src/components/IntlTelInputApp.js [hunk 3]
```diff
@@ -683,25 +683,34 @@ class IntlTelInputApp extends Component {
...
-      value: number,
-    }, () => {
-      if (doNotify) {
-        this.notifyPhoneNumberChange(this.state.value);
-      }
+    this.setState(
+      (prevState) => {
+        if (prevState.showDropDown || prevState.value !== number) {
+          if (doNotify) {
+            this.notifyPhoneNumberChange(number);
```
[the remark this answers is not in the document]
Comment (adrienharnay) [reply src/components/IntlTelInputApp.js]: Previous version would always update the state and then call the callback, now it won't `setState` if there is no need, meaning the callback won't be called. So now we are calling it when it needs to be called

**Comment by adrienharnay** [comment]
Yes of course! Please let me know if you want me to change anything. I've had some troubles wrapping my head around the code before the big `getDerivedStateFromProps`, as some of the code follows this dogma, and some is still based on `componentDidUpdate`

**Comment by superhit0** [comment]
@adrienharnay sorry man, have been too busy with work lately, I will have a look this weekend.

**Pull request closed without merging**